graph/generate.go

```diff
@@ -274,16 +307,60 @@ func valueSourceAncestors(msg *mbp.ValueSource) ([]proto.Message, error) {
 	case *mbp.ValueSource_ConstString:
 		return nil, nil
 	case *mbp.ValueSource_FromInput:
-		return nil, nil
+		ancestor, err := argumentAncestor(e, m.FromInput)
+		if err != nil {
+			return nil, fmt.Errorf("getting ancestors for msg %v failed with error:\n%w", msg, err)
+		}
+		return []proto.Message{ancestor}, nil
 	default:
 		return nil, fmt.Errorf("interpreting whistler message %v failed; type %T not supported", m, m)
 	}
 }
 
+func argumentAncestor(e *env, msg *mbp.ValueSource_InputSource) (proto.Message, error) {
+	index := int(msg.Arg) - 1
+	if index < 0 || index >= len(e.args) {
+		return nil, fmt.Errorf("msg %v requested out-of-bounds argument %v from arguments %v", msg, index, e.args)
+	}
+	return e.args[index], nil
+}
+
 func projectorAncestors(msg *mbp.ProjectorDefinition) []proto.Message {
 	ancestors := make([]proto.Message, len(msg.GetMapping()))
 	for i, mapping := range msg.GetMapping() {
 		ancestors[i] = proto.Message(mapping) // loop over the mappings to convert them to messages
 	}
 	return ancestors
 }
+
+// TODO: In the refactor, the "descendantMsg" is always the projector's valueSource and no switch is needed
+//       This is because extra context is immediately stored in a whistlerNode,
+//       rather than relying on the descendant message for extra context
+// projectorArgs returns the arguments of a projector.
+// Because a ProjectorDefinition message does not have the arguments, it uses
+// the projector's descendant node. This can be a ValueSource if its descendant
+// was a ProjectedValue, or a FieldMapping otherwise.
+func projectorArgs(descendantMsg proto.Message) ([]proto.Message, error) {
+	var source *mbp.ValueSource
+	switch msg := descendantMsg.(type) {
+	case *mbp.FieldMapping:
+		source = msg.GetValueSource()
+	case *mbp.ValueSource:
+		if msg.GetProjector() == "" {
+			return nil, fmt.Errorf("expected ValueSource message {%v} to have a projector set, but it did not", descendantMsg)
+		}
+		source = msg
+	default:
+		return nil, fmt.Errorf("expected message of type FieldMapping or ValueSource, but got %T", descendantMsg)
+	}
+
+	if source.GetSource() == nil {
+		return []proto.Message{}, nil
+	}
+	args := make([]proto.Message, len(source.GetAdditionalArg())+1)
+	args[0] = source
+	for i, arg := range source.GetAdditionalArg() {
```

Comment:
nit: https://golang.org/pkg/builtin/#copy

Reply:
Args has type []proto.Message and source.GetAdditionalArg() has type []*ValueSource. Converting between the two so that copy can be used requires iterating over the ValueSources to convert them one-by-one, so iteration can't be avoided here.